File: backend/app/schemas/shopping.py
from pydantic import BaseModel, Field
from pydantic import BaseModel, Field
from typing import Any, Optional


class ShoppingIntent(BaseModel):
    product_type: str
    budget_max: float | None = None
    budget_min: float | None = None
    currency: str
    use_case: str
    priority_features: list[str] = []
    country: str


class ProductCandidate(BaseModel):
    title: str
    url: str
    snippet: str
    source: str
    price: float | None = None
    currency: str
    is_shopping_site: bool = False
    image: str | None = None
    specs: dict = {}


class ShoppingSearchResponse(BaseModel):
    query: str
    intent: ShoppingIntent
    budget: float | None = None
    currency: str
    count: int
    products: list[ProductCandidate]


class ShoppingCompareRequest(BaseModel):
    query: str = Field(min_length=2, max_length=300)
    top_n: int = Field(default=5, ge=1, le=10)   # 👈 raise limit if needed
    budget: Optional[float] = None
    currency: Optional[str] = None


class ProductEnriched(BaseModel):
    title: str
    url: str
    snippet: str
    source: str
    price: float | None = None
    currency: str
    is_shopping_site: bool = False
    image: str | None = None
    specs: dict = {}
    product_name: str | None = None
    brand: str | None = None
    model: str | None = None
    highlights: list[str] = []
    release_year: int | None = None
    score: int = 0
    score_reasons: list[str] = []
    enrich_error: str | None = None


class ShoppingEnrichedResponse(BaseModel):
    intent: dict[str, Any]
    products: list[dict[str, Any]]
    enriched_count: int = 0
    elapsed_seconds: Optional[float] = None
    

File: frontend/lib/shopping.ts
import api from "./api";

// ═══════════════════════════════════════════════════════════════
// TYPES
// ═══════════════════════════════════════════════════════════════
export type ShoppingIntent = {
  product_type: string;
  budget_max: number | null;
  budget_min: number | null;
  currency: string;
  use_case: string;
  priority_features: string[];
  location?: string | null;
};

export type ProductSpecs = {
  [key: string]: string | number | boolean | null;
};

export type ProductReviews = {
  summary?: string;
  positives?: string[];
  negatives?: string[];
  average?: number;
  count?: number;
};

export type EnrichedProduct = {
  title: string;
  product_name?: string;
  url: string;
  image?: string;
  price: number | null;
  currency: string;
  brand?: string;
  rating?: number;
  reviews_count?: number;
  source?: string;
  site?: string; 
  specs?: ProductSpecs;
  reviews?: ProductReviews;
  score: number;
  score_reasons: string[];
  trust_level?: "high" | "medium" | "low";
};

export type ShoppingCompareResponse = {
  intent: ShoppingIntent;
  products: EnrichedProduct[];
};

export type ShoppingSearchProduct = {
  title: string;
  url: string;
  image?: string;
  price?: number | null;
  currency?: string;
  source?: string;
  site?: string;
};

export type ShoppingSearchResponse = {
  products: ShoppingSearchProduct[];
  total?: number;
};

// ═══════════════════════════════════════════════════════════════
// SEARCH — fast, cheap, no enrichment
// ═══════════════════════════════════════════════════════════════
export const shoppingSearch = async (
  query: string,
  maxResults: number = 5
): Promise<ShoppingSearchResponse> => {
  const res = await api.get("/api/shopping/search", {
    params: { q: query, max_results: maxResults },
  });
  return res.data;
};

// ═══════════════════════════════════════════════════════════════
// COMPARE — deep, slower (fetches + enriches + ranks)
// Uses POST so we can send a larger structured request body
// ═══════════════════════════════════════════════════════════════
export const shoppingCompare = async (
  query: string,
  topN: number = 3,
  options?: {
    budget?: number | null;
    currency?: string | null;
  }
): Promise<ShoppingCompareResponse> => {
  const res = await api.post("/api/shopping/compare", {   // ← confirm endpoint in openapi
    query,
    top_n: topN,
    budget: options?.budget ?? null,
    currency: options?.currency ?? null,
  });
  return res.data;
};